crates/compression/src/compress.rs

```diff
@@ -0,0 +1,223 @@
+            /// Converts the preparation context into a [`CompressCtx`]
+            /// keeping accessed keys to avoid its eviction during compression.
+            /// Initializes the cache evictors from the database, which may fail.
+            pub fn into_compression_context(mut self) -> anyhow::Result<CompressCtx<D>> {
+                Ok(CompressCtx {
+                    $(
+                        $ident: CompressCtxKeyspace {
+                            changes: Default::default(),
+                            cache_evictor: CacheEvictor::new_from_db(&mut self.db, self.accessed_keys.$ident.into())?,
+                        },
+                    )*
+                    db: self.db,
+                })
+            }
+        }
+
+        impl<D> CompressCtx<D> where D: CompressDb {
+            /// Finalizes the compression context, returning the changes to the registry.
+            /// Commits the cache evictor states to the database.
+            fn finalize(mut self) -> anyhow::Result<RegistrationsPerTable> {
+                let mut registrations = RegistrationsPerTable::default();
+                $(
+                    self.$ident.cache_evictor.commit(&mut self.db)?;
+                    for (key, value) in self.$ident.changes.into_iter() {
+                        registrations.$ident.push((key, value));
+                    }
+                )*
+                Ok(registrations)
+            }
+        }
+
+        $(
+            impl<D> CompressibleBy<PrepareCtx<D>> for $type
+            where
+                D: CompressDb
+            {
+                async fn compress_with(
+                    &self,
+                    ctx: &mut PrepareCtx<D>,
+                ) -> anyhow::Result<RegistryKey> {
+                    if *self == <$type>::default() {
+                        return Ok(RegistryKey::ZERO);
+                    }
+                    if let Some(found) = ctx.db.registry_index_lookup(self)? {
+                        ctx.accessed_keys.$ident.insert(found);
+                    }
+                    Ok(RegistryKey::ZERO)
+                }
+            }
+
+            impl<D> CompressibleBy<CompressCtx<D>> for $type
+            where
+                D: CompressDb
+            {
+                async fn compress_with(
+                    &self,
+                    ctx: &mut CompressCtx<D>,
+                ) -> anyhow::Result<RegistryKey> {
+                    if self == &Default::default() {
+                        return Ok(RegistryKey::DEFAULT_VALUE);
+                    }
+                    if let Some(found) = ctx.db.registry_index_lookup(self)? {
+                        return Ok(found);
+                    }
+
+                    let key = ctx.$ident.cache_evictor.next_key();
+                    let old = ctx.$ident.changes.insert(key, self.clone());
+                    assert!(old.is_none(), "Key collision in registry substitution");
```

Comment:
We don't handle it here, and allocate two keys in that case. That's not correct and it's fixed in 40a1a3dc2e.

We would not hit the assert anyway, since that requires that the cache evictor would return the same key twice, which will not happen.